node/lib/util/checkout.js

```diff
@@ -382,18 +403,27 @@ A branch named ${colors.red(name)} already exists.`);
                         },
                     };
                     result.switchBranch = committish;
+                    if (null === result.commit) {
+                        throw new UserError(`\
+Could not resolve ${colors.red(committish)} as a branch or commit.`);
+                    }
                 }
             }
-
-            // If we didn't resolve anything from `committish`, throw an error.
-
-            if (null === result.commit) {
-                throw new UserError(
-         `Could not resolve ${colors.red(committish)} as a branch or commit.`);
+            if (null === annotated) {
+                // If we didn't resolve anything from `committish`, try it
+                // as a file.
+                files.splice(0, 0, committish);
+                // We would now like to resolve from the index, but
+                // libgit2 doesn't support the ":0" shorthand, and our
+                // remaining code all assumes an actual commit, so
+                // we're just going to make one.  The cache-tree ought to
+                // make this relatively cheap.
+
+                annotated = yield makeShadowCommitFromIndex(repo);
```

Comment:
Clever! I'm not sure that this is less complex than allowing `checkoutFiles` to work on an index if `commit` is null, but I don't have a strong opinion.

When we want to support `--ours` and `--theirs`, can we make these options to `makeShadowCommitFromIndex`?

Reply:
I considered something like null, or special-casing ":0" (and :1 &c), but decided this was simpler.  And I think --ours and --theirs will fall out naturally, just as you suggest.